### ISSUE TITLE
Pointers

### ISSUE THREAD

**Issue by Udayv56** [Udayv56/CPP#1]
In C++, a pointer is a variable that holds the memory address of another variable. Pointers are used to indirectly access and manipulate the values stored in memory.

**CODE EXAMPLE 1 -** 
#include <iostream>

using namespace std;

int main()
{
  
    int a=3;
    int* b = &a; //declaration of a pointer to a 
    // & is Address of operator
    // * is dereference operator
    // Variables are given address in RAM where they get stored
    
    cout<<"The address of a is - "<<b<<endl;
    cout<<"The value of a is - "<<*b<<endl;
    cout<<"The address of a is - "<<&a<<endl;
    cout<<"The value of a is - "<<a<<endl;
    
    //pointer to pointer
    int** c = &b;
    cout<<"The address of b is - "<<c<<endl;
    cout<<"The value at address stored in c is - "<<*c<<endl;
    cout<<"The value at address of address of c is - "<<**c<<endl;
    
    return 0;
}

**OUTPUT**-
The address of a is - 0x7fff822473f4
The value of a is - 3
The address of a is - 0x7fff822473f4
The value of a is - 3
The address of b is - 0x7fff822473f8
The value at address stored in c is - 0x7fff822473f4
The value at address of address of c is - 3

**EXPLAINATION**-

In the code provided, the following is happening:

1. int a=3; - a variable a of type int is declared and initialized with the value 3.
2. int* b = &a; - a pointer b of type int* is declared and assigned the memory address of a using the address-of operator &. So, b now holds the address of a.
3. cout<<"The address of a is - "<<b<<endl; - the memory address of a is printed by accessing b.
4. cout<<"The value of a is - "<<*b<<endl; - the value stored at the memory address held by b is printed by dereferencing b using the dereference operator *. Since b holds the address of a, *b will give us the value stored in a.
5. cout<<"The address of a is - "<<&a<<endl; - the memory address of a is printed using the address-of operator &.
6. cout<<"The value of a is - "<<a<<endl; - the value stored in a is printed directly.
7. int** c = &b; - a pointer c of type int** (i.e., a pointer to a pointer) is declared and assigned the memory address of b using the address-of operator &. So, c now holds the address of b.
8. cout<<"The address of b is - "<<c<<endl; - the memory address of b is printed by accessing c.
9. cout<<"The value at address stored in c is - "<<*c<<endl; - the value stored at the memory address held by c is printed by dereferencing c using the dereference operator *. Since c holds the address of b, *c will give us the value stored in b, which is the memory address of a.
10. cout<<"The value at address of address of c is - "<<**c<<endl; - the value stored at the memory address held by *c is printed by dereferencing *c twice using the dereference operator *. Since *c holds the memory address of a, **c will give us the value stored in a.

 So, in short, pointers allow us to indirectly access and manipulate the values stored in memory. They are a powerful tool in C++ programming and are commonly used in low-level programming, data structures, and algorithms.
